File: modules/web/js/ballerina/env/function.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import log from 'log';
import _ from 'lodash';
import EventChannel from 'event_channel';

/**
 * @class Function
 * @augments
 * @param {Object} args - data to create the Function
 * @param {string} args.name - name of function
 * @param {string} args.id - id of function
 * @constructor
 */
class Function extends EventChannel {
    constructor(args) {
        super(args);
        this._name = _.get(args, 'name', '');
        this._id = _.get(args, 'id', '');
        this._parameters = _.get(args, 'parameters', []);
        this._returnParams = _.get(args, 'returnParams', []);
    }

   /**
    * sets the name
    * @param {string} name
    */
    setName(name) {
        var oldName = this._name;
        this._name = name;
        this.trigger("name-modified", name, oldName);
    }

   /**
    * returns the name
    * @returns {string}
    */
    getName() {
        return this._name;
    }

   /**
    * sets the id
    * @param {string} id
    */
    setId(id) {
        this._id = id;
    }

   /**
    * returns the id
    * @returns {string}
    */
    getId() {
        return this._id;
    }

   /**
    * sets the parameters
    * @param [object] parameters
    */
    setParameters(parameters) {
        this._parameters = parameters;
    }

   /**
    * returns the parameters
    * @returns [object]
    */
    getParameters() {
        return this._parameters;
    }

   /**
    * sets the returnParams
    * @param [object] returnParams
    */
    setReturnParams(returnParams) {
        this._returnParams = returnParams;
    }

   /**
    * returns the returnParams
    * @returns [object]
    */
    getReturnParams() {
        return this._returnParams;
    }

   /**
    * sets values from a json object
    * @param {Object} jsonNode
    */
    initFromJson(jsonNode) {
        this.setName(jsonNode.name);
        this.setId(jsonNode.name);
        this.setParameters(jsonNode.parameters);
        this.setReturnParams(jsonNode.returnParams);
    }
}

export default Function;

File: composer/modules/server/services/ballerina-parser/src/main/java/org/ballerinalang/composer/service/ballerina/parser/service/model/lang/Function.java
/*
 * Copyright (c) 2017, WSO2 Inc. (http://wso2.com) All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.ballerinalang.composer.service.ballerina.parser.service.model.lang;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Function.
 */

public class Function {
    private String name = null;

    private Integer packageId = null;

    private String description = null;

    private Integer boundObjectId = null;

    private List<AnnotationAttachment> annotations = new ArrayList<>();

    private List<Parameter> parameters = new ArrayList<Parameter>();

    private List<Parameter> returnParams = new ArrayList<Parameter>();

    private String receiverType = null;

    private Boolean isPublic = false;

    private String fileName = null;

    public Function name(String name) {
        this.name = name;
        return this;
    }

    /**
     * Get name.
     *
     * @return name
     **/
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Function description(String description) {
        this.description = description;
        return this;
    }

    /**
     * Get description.
     *
     * @return description
     **/
    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Function annotations(List<AnnotationAttachment> annotations) {
        this.annotations = annotations;
        return this;
    }

    public Function addAnnotationsItem(AnnotationAttachment annotationsItem) {
        this.annotations.add(annotationsItem);
        return this;
    }

    /**
     * Get annotations.
     *
     * @return annotations
     **/
    public List<AnnotationAttachment> getAnnotations() {
        return annotations;
    }

    public void setAnnotations(List<AnnotationAttachment> annotations) {
        this.annotations = annotations;
    }

    public Function parameters(List<Parameter> parameters) {
        this.parameters = parameters;
        return this;
    }

    public Function addParametersItem(Parameter parametersItem) {
        this.parameters.add(parametersItem);
        return this;
    }

    /**
     * Get parameters.
     *
     * @return parameters
     **/
    public List<Parameter> getParameters() {
        return parameters;
    }

    public void setParameters(List<Parameter> parameters) {
        this.parameters = parameters;
    }

    public Function returnParams(List<Parameter> returnParams) {
        this.returnParams = returnParams;
        return this;
    }

    public Function addReturnParamsItem(Parameter returnParamsItem) {
        this.returnParams.add(returnParamsItem);
        return this;
    }

    /**
     * Get returnParams.
     *
     * @return returnParams
     **/
    public List<Parameter> getReturnParams() {
        return returnParams;
    }

    public void setReturnParams(List<Parameter> returnParams) {
        this.returnParams = returnParams;
    }


    /**
     * Get ReceiverType.
     *
     * @return receiverType
     */
    public String getReceiverType() {
        return receiverType;
    }

    public void setReceiverType(String receiverType) {
        this.receiverType = receiverType;
    }

    /**
     * Returns true if the function is public else false.
     *
     * @return {@link Boolean} whether the function is pubic or not
     */
    public boolean isPublic() {
        return isPublic;
    }

    public void setPublic(Boolean isPublic) {
        this.isPublic = isPublic;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Function function = (Function) o;
        return Objects.equals(this.name, function.name) &&
                Objects.equals(this.description, function.description) &&
                Objects.equals(this.annotations, function.annotations) &&
                Objects.equals(this.parameters, function.parameters) &&
                Objects.equals(this.returnParams, function.returnParams) &&
                Objects.equals(this.receiverType, function.receiverType) &&
                Objects.equals(this.isPublic, function.isPublic);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, description, annotations, parameters, returnParams, receiverType);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("class Function {\n");

        sb.append("    name: ").append(toIndentedString(name)).append("\n");
        sb.append("    description: ").append(toIndentedString(description)).append("\n");
        sb.append("    annotations: ").append(toIndentedString(annotations)).append("\n");
        sb.append("    parameters: ").append(toIndentedString(parameters)).append("\n");
        sb.append("    returnParams: ").append(toIndentedString(returnParams)).append("\n");
        sb.append("    receiverType: ").append(toIndentedString(receiverType)).append("\n");
        sb.append("    isPublic: ").append(toIndentedString(isPublic)).append("\n");
        sb.append("}");
        return sb.toString();
    }

    /**
     * Convert the given object to string with each line indented by 4 spaces (except the first line).
     */
    private String toIndentedString(Object o) {
        if (o == null) {
            return "null";
        }
        return o.toString().replace("\n", "\n    ");
    }

    public String getFileName() {
        return fileName;
    }

    public void setFileName(String fileName) {
        this.fileName = fileName;
    }
}

